hnn_core/network_builder.py

```diff
@@ -425,35 +437,189 @@ def _create_cells_and_feeds(self):
             else:
                 _PC.cell(gid, feed_cell.nrn_netcon)
 
+    def _connect_celltypes(self, src_type, target_type, loc,
+                           receptor, nc_dict, unique=False,
+                           allow_autapses=True):
+        """Connect two cell types for a particular receptor.
+
+        Parameters
+        ----------
+        src_type : str
+            The source cell type
+        target_type : str
+            The target cell type
+        loc : str
+            If 'proximal' or 'distal', the corresponding
+            dendritic sections from Cell.sect_loc['proximal']
+            or Cell.Sect_loc['distal'] are used
+        receptor : str
+            The receptor.
+        nc_dict : dict
+            The connection dictionary containing keys
+            A_delay, A_weight, lamtha, and threshold.
+        unique : bool
+            If True, each target cell gets one "unique" feed.
+            If False, all src_type cells are connected to
+            all target_type cells.
+        allow_autapses : bool
+            If True, allow connecting neuron to itself.
+        """
+        net = self.net
+        connection_name = f'{src_type}_{target_type}_{receptor}'
+        if connection_name not in self.ncs:
+            self.ncs[connection_name] = list()
+        # XXX: self.cells and net._gid_list are not same length
+        # ideally self.cells should be a dict of list
+        for gid_target, target_cell in zip(self.net._gid_list, self.cells):
+            is_target_gid = (gid_target in
+                             self.net.gid_dict[_long_name(target_type)])
+            if _PC.gid_exists(gid_target) and is_target_gid:
```

Comment:
So the MPI bug was caused by referencing a gid that was not assigned to the current node within a parallel context?

Reply:
It's a bit hard to explain and I don't fully understand it, but I can assure you that it works fine :)

The problem is that there isn't necessarily a one-to-one mapping from `self.cells` to gids. That is `self.cells[175]` is not the cell corresponding to `gid=175`. Rather the one-to-one corresponding is between `self.cells` and `self._gid_list`. [This line](https://github.com/jonescompneurolab/hnn-core/pull/164/commits/8d3e5e7269b7e6a20c1298e21b4c6b1c8a62142d#diff-248a9986cd9a62479f5f12b0b576c072L479) was failing. What's not clear to me is why it was working in the non-MPI case ...

Reply:
I trust that it works - I'm just curious at this point. It looks like `self.cells` is [assigned according to the contents of `self._gid_list`](https://github.com/jonescompneurolab/hnn-core/blob/d2addbdea81ad73cab1e72371defbd89cf827a35/hnn_core/network_builder.py#L391). Since the contents of `self._gid_list` is [distinct for each node](https://github.com/jonescompneurolab/hnn-core/blob/d2addbdea81ad73cab1e72371defbd89cf827a35/hnn_core/network_builder.py#L332-L334), I guess it makes sense that `MPIBackend` (which parallelizes across cells) uniquely maps distinct gids and their corresponding cells to different nodes.

Reply:
yeah, I wonder if `self.net._gid_list` should be a `dict` called `gid_dict_node`, so it's easier to manipulate and more transparent what it does